modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java

```diff
@@ -3617,55 +3611,108 @@ public T find() throws IgniteCheckedException {
+            }
+
+            // Release tail if retry is required.
+            if (res == RETRY)
+                releaseTail();
+
+            return res;
         }
 
         /** {@inheritDoc} */
-        @Override boolean notFound(BPlusIO<L> io, long pageAddr, int idx, int lvl) {
-            assert btmLvl >= 0 : btmLvl;
-            assert lvl >= btmLvl : lvl;
+        @Override protected Result finishTail() throws IgniteCheckedException {
+            // An inner node is required for replacement.
+            if (tail.lvl == 0)
+                return NOT_FOUND;
 
-            return lvl == btmLvl;
+            int idx = insertionPoint(tail);
+
+            // Missing row means that current page must be added to tail.
+            if (idx < 0) {
+                idx = fix(idx);
+
+                BPlusInnerIO<L> io = (BPlusInnerIO<L>)tail.io;
+
+                // Release tail in case of broken triangle invariant in locked pages.
+                if (io.getLeft(tail.buf, idx) != tail.down.pageId) {
+                    releaseTail();
+
+                    return RETRY;
+                }
+
+                return NOT_FOUND;
+            }
+
+            assert oldRow == null : "The old row must be set only once.";
+
+            // Insertion index is found. Replace must be performed in both tail top and tail bottom.
+            replaceRowInPage(tail.io, tail.pageId, tail.page, tail.buf, idx);
+
+            // Unlock everything until the leaf, there's no need to hold these locks anymore.
+            while (tail.lvl != 0) {
+                writeUnlockAndClose(tail.pageId, tail.page, tail.buf, null);
+
+                tail = tail.down;
+            }
+
+            // Read old row from the leaf to reduce contention in inner pages.
+            oldRow = needOld ? getRow(tail.io, tail.buf, tail.idx) : (T)Boolean.TRUE;
+
+            // Replace row in the leaf.
+            replaceRowInPage(tail.io, tail.pageId, tail.page, tail.buf, tail.idx);
+
+            finish();
+
+            return FOUND;
         }
 
         /**
+         * Tail page is kept locked after split until insert to the upper level will not be finished. It is needed
+         * because split row will be "in flight" and if we'll release tail, remove on split row may fail.
+         *
          * @param tailId Tail page ID.
          * @param tailPage Tail page pointer.
          * @param tailPageAddr Tail page address
+         * @param io Tail page IO.
+         * @param lvl Tail page level.
          */
-        private void tail(long tailId, long tailPage, long tailPageAddr) {
-            assert (tailId == NULL_PAGE_ID) == (tailPage == NULL_PAGE);
-            assert (tailPage == NULL_PAGE) == (tailPageAddr == NULL_PAGE_ADDRESS);
+        private void setTailForSplit(long tailId, long tailPage, long tailPageAddr, BPlusIO<L> io, int lvl) {
+            assert tailId != 0L && tailPage != 0L && tailPageAddr != 0L;
```

Comment:
why did you remove the constants?

Reply:
1. They were used here and in 2 (I think) more places where I replaced them with "foo = 0L".
2. Every other place in the code uses 0L explicitly. So, we should either used such constants everywhere or not using them at all

Reply:
Is it possible to use them everywhere?

Reply:
That's too many small changes for a bug fix

Reply:
okay =(